### ISSUE TITLE
[Design] Reliable integrations with the full node

### ISSUE THREAD

**Issue by lucas3003** [HathorNetwork/hathor-core#405]
_This design has been moved and updated [here](https://github.com/HathorNetwork/rfcs/blob/master/projects/reliable-integration/0001-high-level-design.md)._

## Problem

Currently, applications that want to interact with the full-node must write its own sync algorithm and handle all use cases (Like reorganization). This algorithm can become very complex and consume several days of development.

There are two alternatives when applications want to interact with the full node:

- Initialize a bidirectional communication via websocket.
- Call the REST APIs.

However, the main concerns with these approaches are:

- When a reorganization happens, the applications must know how to query all the affected txs and update them on their databases.
- There is no guarantee on the order the events happened.
- When the websocket is disconnected, events might be lost and a resync must be executed.

## Solution

To tackle the problems presented above, we must implement a built-in event management system, where all events will be sent in the order they occurred. This system have the following requirements:

1. Detect events that are important to applications. (Look for `Event Types` below).
1. Persist each event and give it an unique incremental ID.
1. Give users an REST API and Websocket connection to query for events

To set-up this system, the user must provide, during the full-node initialization, the `--enable-event-queue` flag.

Due to the necessary flags and the events that can be emitted, we will provide a document on how to understand this new mechanism.

This project, however, does **NOT** intend to eliminate the necessity of a sync algorithm, but to make it much more simpler.

## Out of Scope

These features will not be part of the first phase of this project:

- Sync pause/resume
  - An API to manipulate the sync algorithm.
- Event filter
  - Give users the choice to receive only a subset of events, according to some criteria.
- The `--flush-events` flag. By default, all events are retained. In the future, the user could provide a `--flush-events` flag to enable the flushing of events after each event is sent to the client. 

## Event generation during the full-node cycle

Considering this full-node cycle:

![Full-Node-Cycle drawio](https://user-images.githubusercontent.com/3287486/179076836-ab6ae036-b0c8-449c-a582-2f29ca58ab2b.png)

Where:
- `Load` is the period right after the full-node is started, where the local database is read.
- `Sync` is the period where `Load` is finished and the full-node continuously receive/send txs to other peers until full node is stopped.

By default, the events generated during load will be emitted. If user does not want this behavior, it must provide `--skip-load-events` flag.

## Flow

![Event Flow drawio](https://user-images.githubusercontent.com/3287486/178827445-cc4975f7-3b75-4bc9-a79c-e1042a8a64be.png)

## API

To retrieve the events, an API will be provided:

`GET /event?last_ack_event_id=:last_ack_event_id&size=:size`

The maximum value for `size` will be 1000, to prevent DoS attacks, and the default value will be 100. Those should be configured via settings.

If `last_ack_event_id` is not provided, the first event on the database will be returned.

The result will be an array of events. Each entry will have the data format described on the `Data Format` section below. Also, a `latest_event_id` field will be returned, so the client will know how many events ahead are already available.

## WebSocket

In addition to the REST API, a similar WebSocket API will be available. The client will start the communication also providing the `last_ack_event_id` and an `window_size`. The server will keep the connection open and start sending streaming events to the client.

Each message will be a single event. The server will keep sending events while there are events available and there's `window_size` available. The client can send an ACK message at any time to either update the `ack_event_id` and/or the `window_size`, effectively having full control of the flow of events. Each entry will have the data format described on the `Data Format` section below. Also, a `latest_event_id` field will be returned, so the client will know how many events ahead are already available.

## Storage

To avoid adding new dependencies, the RocksDB, which is already implemented for `tx_storage`, will be used to persist the events. RocksDB is a key-value store. The key will be the event id, and the value will be the whole event. This way, querying a range of keys must be cheap (seek of O(log(n)) and each next is O(1)).

To serialize the data, we will transform the whole event into a JSON object and store as bytes on the RocksDB. A simple `json.decode` will be sufficient to retrieve the encoded data.

## Retention Policy

By default, all events will be stored, given this feature is enable via the `--enable-event-queue` flag. In future phases of the project, a `--flush-events` flag will be implemented to control the retention policy.

## Data Format

All events will have the following structure:

```
{
    full_node_uid: string, // Full node UID, because different full nodes can have different sequences of events
    id: uint64, // Event order
    timestamp: int, // Timestamp in which the event was emitted. This will follow the unix_timestamp format
    type: enum, // One of the event types
    group_id: uint64, // Used to link events. For example, many TX_METADATA_CHANGED will have the same group_id when they belong to the same reorg process
    data: {}, // Variable for event type. Check Event Types section below
}
```

## Data types:

```
tx_input: {
    value: int,
    token_data: int,
    script: string,
    index: int,
    tx_id: bytes,
}
```

```
tx_output: {
    value: int,
    script: bytes,
    token_data: int
}
```

```
token: {
    uid:  string,
    name: string,
    symbol: string
}
```

```
tx: {
    hash: string,
    nonce: int,
    timestamp: long,
    version: int,
    weight: float,
    inputs: tx_input[],
    outputs: tx_output[],
    parents: string[],
    tokens: token[],
    token_name: string,
    token_symbol: string,
    metadata: tx_metadata
}
```

```
spent_outputs: {
    spent_output: spent_output[]
}
```

```
spent_output: {
    index: int,
    tx_ids: string[]
}
```

```
tx_metadata: {
    hash: string,
    spent_outputs: spent_outputs[],
    conflict_with: string[]
    voided_by: string[]
    received_by: int[]
    children: string[]
    twins: string[]
    accumulated_weight: float
    score: float
    first_block: string or null
    height: int
    validation: string
}
```

```
tx_metadata_changed: {
    old_value: tx_metadata,
    new_value: tx_metadata
}
```

```
reorg: {
    reorg_size: int,
    previous_best_block: string, //hash of the block
    new_best_block: string //hash of the block
}
```

## Event Types
- LOAD_STARTED
- LOAD_FINISHED
- NEW_TX_ACCEPTED
- NEW_TX_VOIDED
- NEW_BEST_BLOCK_FOUND
- NEW_ORPHAN_BLOCK_FOUND
- REORG_STARTED
- REORG_FINISHED
- TX_METADATA_CHANGED
- BLOCK_METADATA_CHANGED

### LOAD_STARTED

It will be triggered when the full-node is initializing and the the full node is reading locally from the dabatase, at the same time of `MANAGER_ON_START` Hathor event [here](https://github.com/HathorNetwork/hathor-core/blob/d39fd0d9515c72b5d68b36594294dbd7a155e365/hathor/manager.py#L248). It should have an empty body. 

### LOAD_FINISHED

It will be triggered when the full-node is ready to establish new connections, sync, and exchange transactions, at the same that when the manager state changes to `READY` [here](https://github.com/HathorNetwork/hathor-core/blob/d39fd0d9515c72b5d68b36594294dbd7a155e365/hathor/manager.py#L513). Other events will be triggered ONLY after this one.

### NEW_TX_ACCEPTED

It will be triggered when the transaction is synced, and the consensus algorithm immediately identifies it as an accepted TX that can be placed in the mempool. `tx` datatype is going to be sent. We will reuse the `NETWORK_NEW_TX_ACCEPTED` Hathor Event that is already triggered.

### NEW_TX_VOIDED

It will be triggered inside [this method](https://github.com/HathorNetwork/hathor-core/blob/d39fd0d9515c72b5d68b36594294dbd7a155e365/hathor/consensus.py#L967) when the transaction is received and synced, but the consensus immediately identifies it as a voided TX (As long as a reorg is not in progress). `tx` datatype is going to be sent.

### NEW_BEST_BLOCK_FOUND

It will be triggered when a block is found, received by the full node and immediately identified as a valid block by the consensus algorithm. We will reuse the `NETWORK_NEW_TX_ACCEPTED` Hathor event that is already triggered. In addition, it will trigger ```TX_METADATA_CHANGED``` events for all transactions being directly confirmed by this block (Changing the ```first_block``` information). `tx` datatype is going to  be sent.

### NEW_ORPHAN_BLOCK_FOUND

It will be triggered inside [this method](https://github.com/HathorNetwork/hathor-core/blob/d39fd0d9515c72b5d68b36594294dbd7a155e365/hathor/consensus.py#L443) when a block is found, received by the full node, but immediately identified as a block that does not belong to the best chain. `tx` datatype is going to be sent.

### REORG_STARTED

It will be trigger right above [this line](https://github.com/HathorNetwork/hathor-core/blob/d39fd0d9515c72b5d68b36594294dbd7a155e365/hathor/consensus.py#L293), indicating that the best chain has changed. It will trigger the necessary ```TX_METADATA_CHANGED``` and ```BLOCK_METADATA_CHANGED``` events to void/execute them. `reorg` datatype is going to be sent.

### REORG_FINISHED

It will be triggered right below [this line (outside the for)](https://github.com/HathorNetwork/hathor-core/blob/d39fd0d9515c72b5d68b36594294dbd7a155e365/hathor/consensus.py#L106) if a `REORG_STARTED` had been triggered previously, indicating that the reorg (i.e. a new best chain was found) was completed and all the necessary metadata update was included between ```REORG_STARTED``` and this event.

### TX_METADADATA_CHANGED

Initially, we will trigger this event for two use cases:

- When a best block is found. All transactions will change its ```first_block``` metadata, which will be propagated through this event. This event will be propagated at [this point](https://github.com/HathorNetwork/hathor-core/blob/d39fd0d9515c72b5d68b36594294dbd7a155e365/hathor/consensus.py#L579) 
- When a reorg happens. This can trigger multiple transactions and blocks being changed to voided/executed. This will be detected on `mark_as_voided` functions inside `consensus.py` file (As long as consensus context finds that a reorg is happening). 

Data type `tx_metadata_changed` is going to be sent. Only the affected attributes, along with the hash, will be sent.

## Scenarios

- Two transactions are accepted into the mempool, and a block on the best chain is found to confirm those transactions

    1. NEW_TX_ACCEPTED (Tx 1)
    1. NEW_TX_ACCEPTED (Tx 2)
    1. NEW_BEST_BLOCK_FOUND (Block 1)
    1. TX_METADADATA_CHANGED (Changing the `first_block` of `Tx 1` to `Block 1`)
    1. TX_METADADATA_CHANGED (Changing the `first_block` of `Tx 2` to `Block 1`)

- Two transactions are accepted into the mempool. A block on the best chain is found to confirm those transactions, but a new block on a side chain arrives and becomes the best chain. The transactions are confirmed by this new block.

    1. NEW_TX_ACCEPTED (Tx 1)
    1. NEW_TX_ACCEPTED (Tx 2)
    1. NEW_BEST_BLOCK_FOUND (Block 1)
    1. TX_METADADATA_CHANGED (Changing the `first_block` of `Tx 2` to `Block 1`)
    1. TX_METADADATA_CHANGED (Changing the `first_block` of `Tx 1` to `Block 1`)
    1. REORG_STARTED
    1. NEW_BEST_BLOCK_FOUND (Block 2)
    1. BLOCK_METADATA_CHANGED (Changing the `voided_by` of `Block 1`)
    1. TX_METADATA_CHANGED (Changing the `first_block` of `Tx 1` to `Block 2`)
    1. TX_METADATA_CHANGED (Changing the `first_block` of `Tx 2` to `Block 2`)
    1. REORG_FINISHED

## Integration Tests

We will provide test cases with sequences of events for each scenario. This will help application to integrate with this new mechanism. 

## Task Breakdown and Effort

- [x] Proof of Concept (2 dev-days)
- [x] Low-level design (2 dev-days)
- [x] Change manager to handle the new flag (2 dev-days)
- [x] Make the REORG event be emitted during consensus (2 dev-days)
- [x] Emit event when tx/block is voided (2 dev-days)
- [x] Emit event for metadata changes (2 dev-days)
- [x] Create RocksDB event column family (2 dev-days)
- [x] Implement an event management layer, where events will be detected and persisted on RocksDB (3 dev-days)
- [x] Implement event persistence layer on RocksDB (3 dev-days)
- [x] Implement WebSocket API (2 dev-days)
- [x] Implement `GET /event` REST API  (1 dev-day)
- [ ] Implement `--skip-load-events` flag  (2 dev-days)
- [ ] Doc with user instructions  (1.5 dev-days)
- [ ] Build testing cases for integrations  (2 dev-days)

**Total: 28.5 dev-days**

**Comment by msbrogli** [comment]
Hi @lucas3003 . Thanks for the design! Here are my thoughts:

1) Problem: Another problem happens when the websocket gets disconnected for any reason. In this case, events might be lost so a resync must be executed.

2) API: I feel we should provide a websocket API where a new connection provides the last received id and the websocket stream all events after the given one followed by a real time stream of events. What do you think?

3) Retention policy: Maybe we should give the choice to retain all events. It might be useful when one full node is serving more than one application.

4) Full node uid: As different full nodes will face different sequence of events, we must confirm that the application is connected to the correct full node id before providing the events.

5) Event types: What happens if a transaction in the mempool changes from ACCEPTED to VOIDED? What type of event will be generated in this case?

6) REORG_STARTED: Should we send the reorg size, previous best block, and new best block?

7) Data format: We might add a `group_id` to an event. We can use the same group_id for the TX_METADATA_CHANGED and the REORG_STARTED events, so the application can link both events.

8) API: Should the client provide the last received id? In this case, the API should provide the events just after the given one.

9) API: Will the response have metadata such as "more events available"? Or "latest_event_id"? This might be useful for the application to log how far it is from real time events.

**Comment by lucas3003** [comment]
Thanks for the comments @msbrogli. Here is the answer to your points:

> Problem: Another problem happens when the websocket gets disconnected for any reason. In this case, events might be lost so a resync must be executed.

Included this problem on the design.

> API: I feel we should provide a websocket API where a new connection provides the last received id and the websocket stream all events after the given one followed by a real time stream of events. What do you think?

I agree. I think this might have more value to the clients than the REST API itself. I've included a section about this and updated the task breakdown with this feature.

> Retention policy: Maybe we should give the choice to retain all events. It might be useful when one full node is serving more than one application.

Included in the design a new flag where the user can state that the events must be retained.

> Full node uid: As different full nodes will face different sequence of events, we must confirm that the application is connected to the correct full node id before providing the events.

I included on the data format that each event will return the full node uid.

> Event types: What happens if a transaction in the mempool changes from ACCEPTED to VOIDED? What type of event will be generated in this case?

The idea is to have a event trigger on `mark_as_voided` on the consensus algorithm, so every tx changed from accepted to voided should be catch on this part. A `TX_METADATA_CHANGED` will be generated.

> REORG_STARTED: Should we send the reorg size, previous best block, and new best block?

I have updated the design with a new datatype called `reorg`. So it will not be an empty event anymore, but an event with `size`, a hash of the `previous_best_block` and `new_best_block`.

> Data format: We might add a group_id to an event. We can use the same group_id for the TX_METADATA_CHANGED and the REORG_STARTED events, so the application can link both events.

I added the `group_id` to the event data, so we can group all metadata changes that belongs to the same reorg.

> API: Should the client provide the last received id? In this case, the API should provide the events just after the given one.

I think it makes sense. I included on the API request.

> API: Will the response have metadata such as "more events available"? Or "latest_event_id"? This might be useful for the application to log how far it is from real time events.

I included the `latest_event_id` on the API response.

**Comment by pedroferreira1** [comment]
I really like the design, just some comments:

> GET /event?last_received=:last_received&size=:size

I feel we should have a maximum value here for the `size`, otherwise the user might request a number of events that could DoS the node.

> SYNC_STARTED / SYNC_FINISHED

From the code you linked in the description of these events, I think they should be `LOAD_STARTED` and `LOAD_FINISHED` but I might be wrong. We usually call load the moment that we are loading the transactions/blocks from the local data and sync when we are getting transactions from the p2p network that are old but not in the local database.

> Websocket / API

For the wallet service case I feel the events and API are enough but we've been also discussing using this feature to improve the wallet headless load time. For that we would need a way to get events from a specific wallet but I'm not sure if the full node should be responsible for this filter, or if the wallet should do it locally. This is just a brainstorm and maybe this could be discussed later, at first I feel the wallet could handle this filter to be honest.

**Comment by msbrogli** [comment]
@lucas3003 Thanks for replying. Here are some additional notes:

1) I feel we should have a low level design as well. I guess the POC might help you to do it. As far as I understood, you will add hooks at the consensus algorithm, right?

2) We should include the preparation of a doc with instructions for integration development. It should explain each event type, when they are generated, and explain the concepts needed to develop an integration from scratch. I feel most integrations should have idempotent processors of events.

3) We should include a description of how a new integration would do sync in a fresh start and how it can run a backfill when needed. This should be included in the doc with instructions as well.

4) We might include testing cases for integrations. It would consist of sequences of events for each scenario (e.g., new block found, small reorg, tx changing from executed to voided, etc). This would allow integrations to easily test their code. These tests might send duplicate events since it might happen in a real world case.

5) Will events be triggered during a fresh full node sync? I guess it should. But not during loading since they have already been triggered before.

**Comment by lucas3003** [comment]
@pedroferreira1 comments:

> I feel we should have a maximum value here for the size, otherwise the user might request a number of events that could DoS the node.

I agree. I will limit the `size` to 100. This will not prevent 100% this issue from happening because it will be possible to modify full node's code and set a new size.

> From the code you linked in the description of these events, I think they should be LOAD_STARTED and LOAD_FINISHED but I might be wrong. We usually call load the moment that we are loading the transactions/blocks from the local data and sync when we are getting transactions from the p2p network that are old but not in the local database.

You're right. I was confusing `LOADING` and `SYNCING`. I have changed the events to what you suggested.

**Comment by lucas3003** [comment]
@msbrogli  :

> I feel we should have a low level design as well. I guess the POC might help you to do it. As far as I understood, you will add hooks at the consensus algorithm, right?

You are right. We will be using hooks on the consensus algorithm. I will create the low-level in a separate document.

> We should include the preparation of a doc with instructions for integration development. It should explain each event type, when they are generated, and explain the concepts needed to develop an integration from scratch. I feel most integrations should have idempotent processors of events.

I agree. We must inform users how to use it and what to expect from it. I will add a task for that.

> We should include a description of how a new integration would do sync in a fresh start and how it can run a backfill when needed. This should be included in the doc with instructions as well.

I agree. I will update this design with this point. Also this will be part of the task to write instructions.

> We might include testing cases for integrations. It would consist of sequences of events for each scenario (e.g., new block found, small reorg, tx changing from executed to voided, etc). This would allow integrations to easily test their code. These tests might send duplicate events since it might happen in a real world case.

It is kind of a "simulator", right? I agree this will make easier for applications to test. I will include on this design.

> Will events be triggered during a fresh full node sync? I guess it should. But not during loading since they have already been triggered before.

We actually do not know if during `LOAD` the events have been triggered. For example, if the application uses a snapshot to make loading faster, these events were not emitted.

**Comment by pedroferreira1** [comment]
For me the design is approved.

**Comment by msbrogli** [comment]
This design is approved for me. Thanks, @lucas3003.

**Comment by glevco** [comment]
After starting the implementation, some questions were raised and I discussed design updates with @jansegre.

The main point of discussion is related to this [PR comment](https://github.com/HathorNetwork/hathor-core/pull/457#discussion_r1001197940) by @msbrogli. Previously, the stream of events was being implemented by a `LoopingCall`, that is, every T seconds the server would send a batch o events to the clients, instead of sending each event in real-time.

The problem with sending real-time events is that if the client is not up to date with the latest event, it has to backfill missing events first, because it should process each event in order without having to store them. I discussed some ideas with @jansegre and we concluded that the client could perform the backfill via the REST API (that is, request batches of events in a loop until it is up to date with the stream), and then request the stream of events via WebSocket from the event ID it has backfilled.

The problem here is that there are new events being generated by the server while the client is performing the backfill, so these new events can't be sent to the client in a reactive way (as the events are generated). They would have to be stored in some kind of reactive queue. This queue may never be empty because as the server is popping elements from it, it's also pushing new elements to it. It would be complex to implement this mechanism in a reactive way. I believe this is the reason the first solution was implemented using the `LoopingCall` instead.

I gave more thought into it and came up with another solution: instead of storing events in a queue, the server can react to each new event and either send this new event to client (if the client is up to date), or send a batch of "old" events (if the client is lagged behind). By doing that, it's guaranteed that the client will eventually catch up and it's way less complex to implement than the queue mechanism. Here's an example of the algorithm:

1. Let's say the server is running and has fired and stored a 1000 events.
2. The client connects to the server for the first time and requests 100 events through the REST API
  a. It receives the information that the latest event's ID is `1000`.
  b. It performs REST requests in a loop getting batches of 100 events until it has all 1000 events.
3. The client requests streaming events through the WebSocket API
  a. At this point, let's say the server is already at event ID `1100`.
  b. The client informs the server that the last ID it has received is `1000`.
4. The server generates event `1101` an reacts to it. It checks that the `last_received` by the client is smaller then this new event's ID (`1101 > 1000`)
  a. Instead of sending that new event to the client, it sends a batch of up to, let's say, 100 events (`1001` to `1100`) and updates the `last_received` of the client to `1100`
5. The server generates event `1102`. The same as the last step happens, and it sends a batch of `1001` to `1102`.
6. The server generates event `1103`. At this point the server knows that the last event it sent to the client is the last one that was generated. The client is finally up to date and can receive only the new event `1103`.
7. The server continues to reactively send events one by one to the client, until the client is disconnected and the process restarts.

Of course, numbers like batch sizes can be adjusted.

---

I have edited the original issue description to reflect this, but in high-level.  If this is approved, I'll update the low-level design with the algorithm described above.

I have also made other changes following my discussion with @jansegre, added missing tasks, etc. Please, @msbrogli and @jansegre, could you re-review it? You can check all changes via the edit history, but here's a list of the main updates:

### Changes in the Retention Policy

The implementation of the `--retain-events` flag has been moved out of scope, to be implemented in future phases of this project, both for usability and complexity reasons.

It has been discussed that the base use cases will opt to retain the events, so the default could be to enable this flag. For this reason, the default will be to retain the events, and the future flag has been inverted to `--flush-events`.

Implementation of the `--flush-events` flag would add complexity, because:

1. If events are flushed after being sent to the client, we cannot allow the connection of multiple clients. We have to enforce that only one client can be connected.
2. If the client fails to process an event, it cannot be flushed from the server. The client has to somehow let the server know that it has successfully processed an event.

For the same reason, the `DELETE /event` REST endpoint has been removed.

If we determine this is mandatory in the first phase, we could just create a new task instead of postponing it.

### Changes in the way client uses the REST API vs the WebSocket API

1. The REST API will be used mainly to backfill "old", missed events.
  a. The `size` parameter default of 100 was too small and has been increased to 1000. It should be configured via `HathorSettings`.
2. The WebSocket API will be used mainly to get "new", streaming events. Its high-level description has been updated, and the low-level design will be updated with the description above, if approved.

### Changes in the Task Breakdown

1. A task for the implementation of the `--skip-load-events` has been added. It was described in the original design but wasn't listed.
2. A task for the `GET /event` REST endpoint has been added. It was described in the original design but wasn't listed.

**Comment by msbrogli** [comment]
@glevco Thanks for the reviewing notes. It seems good to me.

**Comment by jansegre** [comment]
✅ Seems good to me too.

**Comment by glevco** [comment]
_This design has been moved and updated [here](https://github.com/HathorNetwork/rfcs/blob/master/projects/reliable-integration/0001-high-level-design.md)._
